Provider crashes on multiple allocations in one pool

Following example code:

```
locals {
  clients_list = toset(["a", "b"])
}

resource "ipam_allocation" "foo" {
  for_each = local.clients_list

  pool_id = ipam_pool.home.id
  size    = 16
}
```

Single change at a time, resulting in success:

```
ipam_allocation.clients["a"]: Creating...
ipam_allocation.clients["a"]: Creation complete after 0s [id=10.0.6.0/24]
```

But multiple resources with `for_each` result in error:

```
Terraform will perform the following actions:

  # ipam_allocation.clients["a"] will be created
  + resource "ipam_allocation" "clients" {
      + cidr    = (known after apply)

      + id      = (known after apply)
      + pool_id = "10.0.0.0/16"
      + size    = 24
    }

  # ipam_allocation.clients["b"] will be created
  + resource "ipam_allocation" "clients" {
      + cidr    = (known after apply)
      + id      = (known after apply)
      + pool_id = "10.0.0.0/16"
      + size    = 24
    }

Plan: 2 to add, 0 to change, 0 to destroy.
```

```
ipam_allocation.clients["a"]: Creating...
ipam_allocation.clients["b"]: Creating...
╷
│ Error: Plugin did not respond
│
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin
│ logs may contain more details.
╵

...

Stack trace from the terraform-provider-ipam_v0.1.1 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xbb0a40]

goroutine 40 [running]:
github.com/agrrh/terraform-provider-ipam/internal/provider.(*AllocationResource).Create(0xc000321110, {0xfeeea0, 0xc000289440
}, {{{{0xff33d8, 0xc00032a450}, {0xd1e320, 0xc00032a3c0}}, {0xff4b68, 0xc0005c9bd0}}, {{{0xff33d8, ...}, ...}, ...}, ...}, ..
.)
        github.com/agrrh/terraform-provider-ipam/internal/provider/allocation_resource.go:134 +0x280

github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).CreateResource(0xc0003a7340, {0xfeeea0, 0xc000289
440}, 0xc000017390, 0xc000017330)
        github.com/hashicorp/terraform-plugin-framework@v1.4.1/internal/fwserver/server_createresource.go:101 +0x578
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ApplyResourceChange(0xc0000174e8?, {0xfeeea0, 0xc
000289440}, 0xc00052e5f0, 0xc0000174e8)
        github.com/hashicorp/terraform-plugin-framework@v1.4.1/internal/fwserver/server_applyresourcechange.go:57 +0x4a5
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ApplyResourceChange(0xc0003a7340, {0xfeeea0?,
 0xc000289170?}, 0xc00052e550)
        github.com/hashicorp/terraform-plugin-framework@v1.4.1/internal/proto6server/server_applyresourcechange.go:55 +0x3ff
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0xc000226f00, {0xfeeea0?, 0xc00028
8990?}, 0xc000194a80)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov6/tf6server/server.go:857 +0x56a
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0xe17d40?, 0xc00
0226f00}, {0xfeeea0, 0xc000288990}, 0xc000194a10, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:467 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00027c5a0, {0xff3da0, 0xc0001031e0}, 0xc0001e6ea0, 0xc0003c5f80, 0x161135
8, 0x0)
        google.golang.org/grpc@v1.58.3/server.go:1374 +0xde7
google.golang.org/grpc.(*Server).handleStream(0xc00027c5a0, {0xff3da0, 0xc0001031e0}, 0xc0001e6ea0, 0x0)
        google.golang.org/grpc@v1.58.3/server.go:1751 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        google.golang.org/grpc@v1.58.3/server.go:986 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 71
        google.golang.org/grpc@v1.58.3/server.go:997 +0x145

Error: The terraform-provider-ipam_v0.1.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

--- a/GNUmakefile
+++ b/GNUmakefile
@@ -4,3 +4,6 @@ default: testacc
 .PHONY: testacc
 testacc:
 	TF_ACC=1 go test ./... -v $(TESTARGS) -timeout 120m
+
+install:
+	go install -v

--- a/internal/provider/allocation_resource.go
+++ b/internal/provider/allocation_resource.go
@@ -110,6 +110,9 @@ func (r *AllocationResource) Configure(ctx context.Context, req resource.Configu
 func (r *AllocationResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
 	var data AllocationResourceModel
 
+	mutex.Lock()
+	defer mutex.Unlock()
+
 	// Read Terraform plan data into the model
 	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
 
@@ -145,6 +148,9 @@ func (r *AllocationResource) Create(ctx context.Context, req resource.CreateRequ
 func (r *AllocationResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
 	var data AllocationResourceModel
 
+	mutex.RLock()
+	defer mutex.RUnlock()
+
 	// Read Terraform prior state data into the model
 	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
 
@@ -171,6 +177,9 @@ func (r *AllocationResource) Read(ctx context.Context, req resource.ReadRequest,
 func (r *AllocationResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
 	var data AllocationResourceModel
 
+	mutex.Lock()
+	defer mutex.Unlock()
+
 	// Read Terraform plan data into the model
 	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
 
@@ -187,6 +196,9 @@ func (r *AllocationResource) Update(ctx context.Context, req resource.UpdateRequ
 
 func (r *AllocationResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
 	var data AllocationResourceModel
+
+	mutex.Lock()
+	defer mutex.Unlock()
 
 	// Read Terraform prior state data into the model
 	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)

--- a/internal/provider/pool_resource.go
+++ b/internal/provider/pool_resource.go
@@ -88,6 +88,9 @@ func (r *PoolResource) Configure(ctx context.Context, req resource.ConfigureRequ
 func (r *PoolResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
 	var data PoolResourceModel
 
+	mutex.Lock()
+	defer mutex.Unlock()
+
 	// Read Terraform plan data into the model
 	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
 
@@ -122,6 +125,9 @@ func (r *PoolResource) Create(ctx context.Context, req resource.CreateRequest, r
 func (r *PoolResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
 	var data PoolResourceModel
 
+	mutex.RLock()
+	defer mutex.RUnlock()
+
 	// Read Terraform prior state data into the model
 	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
 
@@ -145,6 +151,9 @@ func (r *PoolResource) Read(ctx context.Context, req resource.ReadRequest, resp 
 func (r *PoolResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
 	var data PoolResourceModel
 
+	mutex.Lock()
+	defer mutex.Unlock()
+
 	// Read Terraform plan data into the model
 	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
 
@@ -161,6 +170,9 @@ func (r *PoolResource) Update(ctx context.Context, req resource.UpdateRequest, r
 
 func (r *PoolResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
 	var data PoolResourceModel
+
+	mutex.Lock()
+	defer mutex.Unlock()
 
 	// Read Terraform prior state data into the model
 	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)

--- a/internal/provider/provider.go
+++ b/internal/provider/provider.go
@@ -5,6 +5,7 @@ package provider
 
 import (
 	"context"
+	"sync"
 
 	"github.com/hashicorp/terraform-plugin-framework/datasource"
 	"github.com/hashicorp/terraform-plugin-framework/provider"
@@ -17,6 +18,9 @@ import (
 
 // Ensure IPAMProvider satisfies various provider interfaces.
 var _ provider.Provider = &IPAMProvider{}
+
+// Global storage-access lock to use within this plugin
+var mutex sync.RWMutex
 
 // IPAMProvider defines the provider implementation.
 type IPAMProvider struct {

